hc-dpki/src/keypair.rs

```diff
@@ -0,0 +1,528 @@
+use crate::{
+    bundle,
+    holochain_sodium::{aead, kx, random::random_secbuf, secbuf::SecBuf, sign},
+    util,
+};
+use holochain_core_types::{agent::KeyBuffer, error::HolochainError};
+use rustc_serialize::json;
+use std::str;
+
+pub struct Keypair {
+    pub pub_keys: String,
+    pub sign_priv: SecBuf,
+    pub enc_priv: SecBuf,
+}
+
+pub const SEEDSIZE: usize = 32 as usize;
+
+impl Keypair {
+    /// derive the pairs from a 32 byte seed buffer
+    ///  
+    /// @param {SecBuf} seed - the seed buffer
+    pub fn new_from_seed(seed: &mut SecBuf) -> Self {
+        let mut seed = seed;
+        let mut sign_public_key = SecBuf::with_insecure(sign::PUBLICKEYBYTES);
+        let mut sign_secret_key = SecBuf::with_secure(sign::SECRETKEYBYTES);
+        let mut enc_public_key = SecBuf::with_insecure(kx::PUBLICKEYBYTES);
+        let mut enc_secret_key = SecBuf::with_secure(kx::SECRETKEYBYTES);
+
+        sign::seed_keypair(&mut sign_public_key, &mut sign_secret_key, &mut seed).unwrap();
+        kx::seed_keypair(&mut seed, &mut enc_public_key, &mut enc_secret_key).unwrap();
+
+        Keypair {
+            pub_keys: util::encode_id(&mut sign_public_key, &mut enc_public_key),
+            sign_priv: sign_secret_key,
+            enc_priv: enc_secret_key,
+        }
+    }
+
+    /// get the keypair identifier string
+    ///
+    /// @return {SecBuf}
+    pub fn get_id(&mut self) -> String {
+        return self.pub_keys.clone();
+    }
+
+    /// generate an encrypted persistence bundle
+    ///
+    /// @param {SecBuf} passphrase - the encryption passphrase
+    ///
+    /// @param {string} hint - additional info / description for the bundle
+    pub fn get_bundle(
+        &mut self,
+        passphrase: &mut SecBuf,
+        hint: String,
+    ) -> Result<bundle::KeyBundle, HolochainError> {
+        let mut passphrase = passphrase;
+        let bundle_type: String = "hcKeypair".to_string();
+        let skk = KeyBuffer::with_corrected(&self.pub_keys)?;
+        let sk = skk.get_sig() as &[u8];
+        let ekk = KeyBuffer::with_corrected(&self.pub_keys)?;
```

Comment:
Should only do the `with_corrected` once, skk and ekk should have identical data in them